File: windows/src/GoalMaker.Infrastructure/Updates/EcdsaSignatureVerifier.cs
using System.Security.Cryptography;
using GoalMaker.Core.Updates;

namespace GoalMaker.Infrastructure.Updates;

/// <summary>
/// ECDSA P-256 with SHA-256 over the exact manifest bytes; the signature is DER, base64-encoded
/// (ADR 0004). Uses only the platform crypto.
/// </summary>
public sealed class EcdsaSignatureVerifier : ISignatureVerifier, IDisposable
{
    private readonly ECDsa key;

    public EcdsaSignatureVerifier(string publicKeyBase64)
    {
        key = ECDsa.Create();
        key.ImportSubjectPublicKeyInfo(Convert.FromBase64String(publicKeyBase64), out _);
    }

    public bool Verify(ReadOnlySpan<byte> data, string signatureBase64)
    {
        byte[] signature;
        try
        {
            signature = Convert.FromBase64String(signatureBase64.Trim());
        }
        catch (FormatException)
        {
            return false;
        }

        if (signature.Length == 0)
        {
            return false;
        }

        try
        {
            return key.VerifyData(data, signature, HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence);
        }
        catch (CryptographicException)
        {
            return false;
        }
    }

    public void Dispose() => key.Dispose();
}

File: android/app/src/main/kotlin/com/goalmaker/app/data/update/EcdsaSignatureVerifier.kt
package com.goalmaker.app.data.update

import com.goalmaker.app.application.update.SignatureVerifier
import java.security.GeneralSecurityException
import java.security.KeyFactory
import java.security.PublicKey
import java.security.Signature
import java.security.spec.X509EncodedKeySpec
import java.util.Base64

/**
 * ECDSA P-256 with SHA-256 over the exact manifest bytes; the signature is DER, base64-encoded
 * (ADR 0004). Uses only platform crypto, available on every supported Android version.
 */
class EcdsaSignatureVerifier(publicKeyBase64: String) : SignatureVerifier {

    private val publicKey: PublicKey =
        KeyFactory.getInstance("EC").generatePublic(X509EncodedKeySpec(Base64.getDecoder().decode(publicKeyBase64)))

    override fun verify(data: ByteArray, signatureBase64: String): Boolean {
        val signature = try {
            Base64.getDecoder().decode(signatureBase64.trim())
        } catch (_: IllegalArgumentException) {
            return false
        }
        if (signature.isEmpty()) return false
        return try {
            Signature.getInstance("SHA256withECDSA").run {
                initVerify(publicKey)
                update(data)
                verify(signature)
            }
        } catch (_: GeneralSecurityException) {
            false
        }
    }
}
